zebROS_ws/src/ctre_interfaces/src/candle_state_interface.cpp

```diff
@@ -0,0 +1,128 @@
+#include "ctre_interfaces/candle_state_interface.h"
+
+namespace hardware_interface {
+namespace candle {
+
+CANdleColour::CANdleColour(int red, int green, int blue, int white) {
+    this->red = red;
+    this->green = green;
+    this->blue = blue;
+    this->white = white;
+}
+CANdleColour::CANdleColour() {}
+bool CANdleColour::operator!=(const CANdleColour& rhs) {
+    return !(this->red == rhs.red && this->green == rhs.green && this->blue == rhs.blue && this->white == rhs.white);
+}
+bool CANdleColour::operator==(const CANdleColour& rhs) {
+    return (
+        this->red == rhs.red &&
+        this->green == rhs.green &&
+        this->blue == rhs.blue &&
+        this->white == rhs.white
+    );
+}
+
+CANdleAnimation::CANdleAnimation(double speed, int start, int count, CANdleAnimationType type, double brightness, bool reversed, double param4, double param5) {
+    this->speed = speed;
+    this->start = start;
+    this->count = count;
+    this->type = type;
+    this->class_type = CANdleAnimationClass::BaseStandard;
+    this->brightness = brightness;
+    this->reversed = reversed;
+    this->param4 = param4;
+    this->param5 = param5;
+}
+CANdleAnimation::CANdleAnimation(double speed, int start, int count, CANdleAnimationType type, int red, int green, int blue, int white, int direction) {
+    this->speed = speed;
+    this->start = start;
+    this->count = count;
+    this->type = type;
+    this->class_type = CANdleAnimationClass::BaseTwo;
+    this->colour = CANdleColour(red, green, blue, white);
+    this->direction = direction;
+}
+CANdleAnimation::CANdleAnimation() {}
+
+bool CANdleAnimation::operator==(const CANdleAnimation& rhs) {
+    if (this->class_type == rhs.class_type) {
+        if (this->class_type == CANdleAnimationClass::BaseStandard) {
+            return (
+                (this->speed == rhs.speed) &&
+                (this->start == rhs.start) &&
+                (this->count == rhs.count) &&
+                (this->brightness == rhs.brightness) &&
+                (this->param4 == rhs.param4) &&
+                (this->param5 == rhs.param5) &&
+                (this->type == rhs.type)
+            );
+        } else {
+            return (
+                (this->speed == rhs.speed) &&
+                (this->start == rhs.start) &&
+                (this->count == rhs.count) &&
+                (this->colour == rhs.colour) &&
+                (this->direction == rhs.direction) &&
+                (this->type == rhs.type)
+            );
+        }
+    } else {
+        return false;
+    }
+}
+bool CANdleAnimation::operator!=(const CANdleAnimation& rhs) {
+    return !(this->operator==(rhs));
+}
+
+CANdleHWState::CANdleHWState(int id) :
+    device_id{id},
+    brightness{1},
+    leds(8),
+    animation()
+{
+    for (int i = 0; i < 8; i++) {
+        this->leds[i] = CANdleColour(255, 255, 255, 255);
+    }
+}
+
+void CANdleHWState::setLED(int id, CANdleColour led) {
+    this->leds[id] = led;
```

Comment:
What happens if code tries to set an LED on the strip (ids 9 and above)?

Reply:
Tested this on the robot-in-a-box, I got animations and normal colours working on a neopixel strip connected to the CANdle. I'm assuming that's what you're talking about so I'll mark this as resolved

Reply:
I was more concerned with the potential for writing past the end of the size of leds, not sure how resizing works for that vector

Reply:
well, it works regardless, lol. I can write to those LEDs